### PR TITLE
Upgrade oz

### PR DESCRIPTION
**Pull request by princetonbishop**
# Description
What does this PR solve?

Fixes # (issue)

# Checklist
- [ ] Code follows the style guide
- [ ] I have performed a self-review of my own code
- [ ] New and existing tests pass locally
- [ ] This PR is targeting the correct branch

### CODE REVIEW

File: protocol/contracts/governance/ElderElection.sol [hunk 1]
```diff
@@ -121,7 +118,7 @@ contract ElderElection is AccessControl {
             hash(req)
         ));
 
-        (address signer, ECDSA.RecoverError err) = ECDSA.tryRecover(digest, signature);
+        (address signer, ECDSA.RecoverError err, bytes32 sigLength) = ECDSA.tryRecover(digest, signature);
```
Comment (frontier159) [inline protocol/contracts/governance/ElderElection.sol:121]: `sigLength` is unused, can remove to get rid of the compiler warning.

File: protocol/contracts/core/RebasingERC20.sol [hunk 2]
```diff
@@ -46,6 +46,43 @@ abstract contract RebasingERC20 is ERC20 {
         return toTokenAmount(shareBalanceOf[account]);
     }
 
+    /**
```
Comment (frontier159) [inline protocol/contracts/core/RebasingERC20.sol:49]: OZ recommends override the `_update()` function for all of these, which imo is better than introducing new non-virtual functions below

File: protocol/contracts/v2/strategies/AbstractStrategy.sol [hunk 3]
```diff
@@ -198,7 +198,7 @@ abstract contract AbstractStrategy is ITempleStrategy, TempleElevatedAccess {
     function _setTokenAllowance(IERC20 token, address spender, uint256 amount) internal {
         if (amount == token.allowance(address(this), spender)) return;
 
-        token.safeApprove(spender, 0);
+        token.approve(spender, 0);
```
Comment (frontier159) [inline protocol/contracts/v2/strategies/AbstractStrategy.sol:201]: can use `forceApprove()` here

File: protocol/contracts/v2/strategies/DsrBaseStrategy.sol [hunk 4]
```diff
@@ -57,7 +57,8 @@ contract DsrBaseStrategy is AbstractStrategy, ITempleBaseStrategy {
         vat.hope(address(daiJoin));
         vat.hope(address(pot));
         daiToken = IERC20(_daiToken);
-        daiToken.safeApprove(address(daiJoin), type(uint256).max);
+        daiToken.approve(address(daiJoin), 0);
```
Comment (frontier159) [inline protocol/contracts/v2/strategies/DsrBaseStrategy.sol:60]: forceApprove

File: protocol/contracts/v2/templeLineOfCredit/TempleLineOfCredit.sol [hunk 5]
```diff
@@ -391,15 +391,15 @@ contract TempleLineOfCredit is ITempleLineOfCredit, TempleElevatedAccess {
...
+            daiToken.approve(previousTrv, 0);
         }
 
         address _trv = address(tlcStrategy.treasuryReservesVault());
         treasuryReservesVault = ITreasuryReservesVault(_trv);
 
         // Set max allowance on the new TRV
         {
-            daiToken.safeApprove(_trv, 0);
+            daiToken.approve(_trv, 0);
```
Comment (frontier159) [inline protocol/contracts/v2/templeLineOfCredit/TempleLineOfCredit.sol:402]: forceApprove

File: protocol/contracts/zaps/GenericZap.sol [hunk 6]
```diff
@@ -37,7 +37,7 @@ contract GenericZap is ZapBase {
 
   constructor (
```
Comment (frontier159) [inline protocol/contracts/zaps/GenericZap.sol:38]: I thought we removed these from the repo :-)

File: protocol/test/core/vault-proxy-tests.ts [hunk 7]
```diff
@@ -231,7 +237,8 @@ describe("Vault Proxy", async () => {
   })
 
   it("Can proxy deposit for any Vault", async() => {
-    await TEMPLE.connect(alan).increaseAllowance(VAULT_PROXY.address, toAtto(1000));
+    await TEMPLE.connect(alan).approve(VAULT_PROXY.address, 0);
```
Comment (frontier159) [inline protocol/test/core/vault-proxy-tests.ts:240]: I'm surprised we need to approve(0) first for the Temple ERC20. I thought it was only USDT that needed it as it's custom.

Has something changed in ERC20 where this is required now

File: protocol/test/amm_test.ts [hunk 8]
```diff
@@ -243,7 +247,8 @@ describe("AMM", async () => {
   describe("Update Treasury", async() => {
 
     it("non-owner reverts", async () => {
-        await shouldThrow(templeRouter.connect(alan).setTreasury(await ben.getAddress()), /Ownable: caller is not the owner/);
+        await shouldThrow(templeRouter.connect(alan).setTreasury(await ben.getAddress()), /OwnableUnauthorizedAccount/);
```
Comment (frontier159) [inline protocol/test/amm_test.ts:250]: Reason you're not checking for custom errors in these too?

File: protocol/contracts/zaps/GenericZap.sol (hunk 6 above)
Comment (princetonbishop) [reply protocol/contracts/zaps/GenericZap.sol:38]: :-)

File: protocol/test/core/vault-proxy-tests.ts (hunk 7 above)
Comment (princetonbishop) [reply protocol/test/core/vault-proxy-tests.ts:240]: after checking the ERC20 contract inherited, it's fine to use one approve for temple token

File: protocol/contracts/v2/strategies/AbstractStrategy.sol [hunk 9]
```diff
@@ -198,7 +198,7 @@ abstract contract AbstractStrategy is ITempleStrategy, TempleElevatedAccess {
...
         if (amount > 0) {
```
Comment (frontier159) [inline protocol/contracts/v2/strategies/AbstractStrategy.sol:202]: Think you missed using forceApprove on this one?

Comment (princetonbishop) [reply protocol/contracts/v2/strategies/AbstractStrategy.sol:202]: good catch!

**Review by frontier159 — APPROVED** [review]

**Pull request merged**